### DISCUSSION TITLE
Branch Prediction Traces

### DISCUSSION THREAD

**Discussion by SteliosKaragiorgis** [ChampSim/ChampSim#443] [General]
Hello. I want to run Branch traces from **Championship Branch Prediction (CBP-5)** .  I have the traces from this google drive: https://drive.google.com/drive/folders/1VAdmqdOEFLvnRKkQQidxvGJA_C6S2RWo . My only problem now is that these traces are not compatible with champsim and I need to convert it from **.bt9.trace** to **.champsimtrace** like the https://github.com/ChampSim/ChampSim/blob/master/tracer/cvp_converter/cvp2champsim.cc (which converts the cvp traces to champsim trace) . Is it any way that I can run these traces **(.bt9.trace)** on the champsim? Because I want to run my branch predictors

Thanks in advance

**Comment by useredsa** [comment]
Hello @SteliosKaragiorgis,

You cannot easily translate BT9 traces to Champsim traces because BT9 traces only contain information about branch instructions. With BT9 traces you can only get microarchitecture-agnostic measures measures (like MPKI, misprediction rate, etc.). Champsim, on the other hand, is a microarchitecture simulator (a great one). For example, it can give you overall performance improvement metrics. But it needs more information than just that the branch instructions seen during the execution.

If you are still interested in running the traces and checking the MPKI of you branch predictors I suggest you check out [MBPlib](https://github.com/useredsa/MBPlib) (disclaimer: I am the author). MBPlib is similar to the simulator of the CBP5 in that it only takes into account branch instructions and as such gives only microarchitecture-agnostic measures. Thanks to this it is also much faster than a simulator that iterates over all instructions. In the repo you can download the CBP5 traces translated to the format used by the simulator. Let me know if it serves you :).

Best regards,

**Reply by SteliosKaragiorgis** [reply]
Hello @useredsa 
Thank you very much for your answer. I will check it!!
 Because i have it for a project i need to show the results on the ChampSim. Do you know any tool where i can open and read ChampSim trace files? I managed to read bt9 trace files and now i need to read a champsim trace file. I think I found a way to read the .champsimtrace file but im not sure if it is correct

**Reply by useredsa** [reply]
I'm not sure I understand your question. If you want to simulate against champsim traces then one option is of course using ChampSim, and this will also give you performance measurements. However, if you want faster simulations because you need to iterate quickly through versions or test multiple configurations, or you would prefer to use a single simulator for your current task, what you can do is to translate your champsim traces to SBBT, the format used by MBPlib. There's a conversor [here](https://github.com/useredsa/champsimtrace_to_sbbt).

**Reply by SteliosKaragiorgis** [reply]
For my thesis project i need to do a branch predictor. So i need to show the results on the ChampSim. But since the ChampSim doesnt have traces for branch prediction then i need to take the traces from CBP-5 and convert them into the ChampSim trace format so i can run my predictor on the ChampSim with branch prediction traces. 
I can open the link of the conversor

Thank you!!

**Reply by useredsa** [reply]
But, as I told you, you cannot do that. The CBP5 traces don't have the information that should be in a ChampSim trace, there's no workaround. The conversor I sent you is to pass from Champsim to SBBT, not the other way around. If your requirements are developing a branch predictor and testing it on the CBP5 traces then you can do that without ChampSim. If you need to use ChampSim, then you cannot use the CBP5 traces. You need to choose.

**Reply by SteliosKaragiorgis** [reply]
I need to do that on ChampSim but because i will do branch predictor i cannot use the given traces since they are for Data Prefetching and Cache Replacement. So i need to run traces for branch prediction but on the ChampSim.

Also I want to read the .champsimtraces files so i can open the trace files. I dont know if there is a tool on the internet or any parser. 
Thanks for the help!!

**Reply by useredsa** [reply]
You can do branch prediction on any trace. Only that if the traces are for the Data Prefetching Championship maybe they do not contain regions of the code with hard to predict branches.

> So i need to run traces for branch prediction but on the ChampSim.

Again, you cannot run BT9 traces from the Championship Branch Prediction in ChampSim. You will have to look for an alternative.

> Also I want to read the .champsimtraces files so i can open the trace files. I dont know if there is a tool on the internet or any parser.

You can look at the source code for the tracereader of ChampSim. It's not very long. From there you can code a tool that will do what you want.
